### PR TITLE
EMTF Emulator Update to Add Displaced Muon pT & Dxy Assignment

### PR DESCRIPTION
**Pull request by eyigitba**
#### PR description:

This PR updates EMTF emulator by adding a new class (`PtAssignmentEngineDxy`) that handles vertex unconstrained pT and Dxy assignment. The changes should not affect anything for prompt muons. The main changes are:

- Added `PtAssignmentEngineDxy` and necessary changes to the EMTF algorithm to support it.
- Added etters and getters for vertex unconstrained pT and Dxy to `EMTFTrack`
- Added a binary file in `L1Trigger/L1TMuonEndCap/data/emtfpp_tf_graphs`

This PR will also be backported to 11_2_X.

#### PR validation:

Validated with:

- scram b runtests
- runTheMatrix.py -l limited -i all --ibeos

Notifications: @jiafulow

### CODE REVIEW

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
Hi @eyigitba 

A PR for the needed data in externals has been made https://github.com/cms-data/L1Trigger-L1TMuon/pull/19
Please adjust this PR to reflect the change I made in the location of data done in commit 9e12f1a 
Thanks.

**Comment by eyigitba** [comment]
Hi @rekovic 

Thanks for the notification. I changed the path to the new one.

**Comment by qliphy** [comment]
test parameters:
pull_request = cms-data/L1Trigger-L1TMuon#19

**Comment by qliphy** [comment]
please test

**Comment by eyigitba** [comment]
I don't understand the RelVals-INPUT fails. They seem to be not related to the changes. Is there something I need to do about this?

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
+1

**Comment by silviodonato** [comment]
+1

**Comment by smuzaffar** [comment]
this also needed cms-data/L1Trigger-L1TMuon#19 to go in . I have merged it and started an IB at 8h00

**Comment by makortel** [comment]
Seems that this PR and https://github.com/cms-data/L1Trigger-L1TMuon/pull/19 are causing segfaults in the IB, see https://github.com/cms-sw/cmsdist/issues/6618 for more information.

**Comment by eyigitba** [comment]
Hi @makortel , thanks for letting us know. I'm currently investigating the issue using the workflows from https://cms-sw.github.io/relvalLogDetail.html#slc7_amd64_gcc900;CMSSW_11_3_X_2021-02-03-0800

It seems like the segfaults go away if I switch to single thread. Using 4 or 8 threads causes segfaults, but not every time I run the cmsDriver command. I'll investigate further to see where it might come from. In the meanwhile if any experts have other ideas I can try them as well.

File: L1Trigger/L1TMuonEndCap/src/PtAssignmentEngineDxy.cc [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+  feature = {{x_dphi[0],      x_dphi[1],      x_dphi[2],      x_dphi[3],      x_dphi[4],    x_dphi[5],
+              x_dtheta[0],    x_dtheta[1],    x_dtheta[2],    x_dtheta[3],    x_dtheta[4],  x_dtheta[5],
+              x_bend_emtf[0], x_bend_emtf[1], x_bend_emtf[2], x_bend_emtf[3], x_fr_emtf[0], x_trk_theta[0],
+              x_me11ring[0],  x_rpcbit[0],    x_rpcbit[1],    x_rpcbit[2],    x_rpcbit[3]}};
+  return;
+}
+
+void PtAssignmentEngineDxy::call_tensorflow_dxy(const emtf::Feature& feature, emtf::Prediction& prediction) const {
+  static tensorflow::Tensor input(tensorflow::DT_FLOAT, {1, emtf::NUM_FEATURES});
+  static std::vector<tensorflow::Tensor> outputs;
```
Comment (makortel) [inline L1Trigger/L1TMuonEndCap/src/PtAssignmentEngineDxy.cc:243]: These are not thread safe.

Comment (eyigitba) [reply L1Trigger/L1TMuonEndCap/src/PtAssignmentEngineDxy.cc:243]: I just removed those 'static' lines and I'm now testing again. Thanks for the input!

File: L1Trigger/L1TMuonEndCap/src/PtAssignmentEngineDxy.cc [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+  // Mimic Phase-1 EMTF input calculations
+  // 6 delta Phis: S1-S2, S1-S3, S1-S4, S2-S3, S2-S4, S3-S4
+  // 6 delta Thetas: S1-S2, S1-S3, S1-S4, S2-S3, S2-S4, S3-S4
+  // 4 bends : set to zero if no CSC hit and thus RPC hit is used
+  // 1 FR bit: for ME1 only
+  // 1 Ring bit: for ME1 only
+  // 1 track Theta taken from stub coordinate in ME2, ME3, ME4 (in this priority)
+  // 4 RPC bits indicating if ME or RE hit was used in each station (S1, S2, S3, S4)
+  // Total: 23 variables
+  static std::array<float, 6> x_dphi;
```
Comment (Dr15Jones) [inline L1Trigger/L1TMuonEndCap/src/PtAssignmentEngineDxy.cc:63]: These static arrays are also not thread safe

**Comment by Dr15Jones** [comment]
The output of the static analyzer did show the thread safety problems with the statics plus some other items:

https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-6d2637/12508/llvm-analysis/

**Comment by eyigitba** [comment]
@Dr15Jones @makortel thanks for all the help! 

I'll submit a new PR fixing thread safety issues with static calls. I don't want to touch the other parts that are not relevant to this PR for now if that's OK.

**Comment by eyigitba** [comment]
New PR is submitted here: #32813

**Comment by eyigitba** [comment]
> The output of the static analyzer did show the thread safety problems with the statics plus some other items:
> 
> https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-6d2637/12508/llvm-analysis/

@Dr15Jones is it possible to use this tool for individual packages to test our changes?

**Comment by Dr15Jones** [comment]
>The output of the static analyzer did show the thread safety problems with the statics plus some other items:
>https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-6d2637/12508/llvm-analysis/

@smuzaffar can one run the static analyzer standalone?

**Comment by smuzaffar** [comment]
one can run the following to run the static analyzer checker. 
```
export USER_CXXFLAGS="-DEDM_ML_DEBUG -w"
export USER_LLVM_CHECKERS="-enable-checker threadsafety -enable-checker optional.ClassChecker -enable-checker cms -disable-checker cms.FunctionDumper"
scram b -k -j $(nproc) checker 
```

**Comment by eyigitba** [comment]
Actually I figured it out.  `-enable-checker optional.ClassChecker` was causing the crashes for some reason. It runs just fine without that option. Thanks a lot again!

**Comment by smuzaffar** [comment]
@eyigitba , try checking out `Utilities/StaticAnalyzers` package in your dev area

**Comment by eyigitba** [comment]
@smuzaffar checking out the package and using the full options also worked. thanks a lot again!

**Pull request merged**